visualizer/main.py

```diff
@@ -0,0 +1,399 @@
+            "dark_yellow_1": (0.94, 0.76, 0.2),
+            "dark_green_1": (0.42, 0.66, 0.31),
+            "dark_cornflower_blue 1": (0.24, 0.5, 0.85),
+            "dark_red_1": (0.8, 0, 0),
+            'dark_purple_1': (0.4, 0.31, 0.65),
+            'dark_magenta_1': (0.65, 0.3, 0.47),
+            'dark_cyan_1': (0.27, 0.51, 0.56),
+            'dark_orange_1': (0.9, 0.57, 0.22),
+        }
+
+        self.d_c_dict = {
+            "dark_yellow_3": (0.49, 0.37, 0),
+            "dark_green_3": (0.15, 0.31, 0.07),
+            "dark_cornflower_blue 3": (0.11, 0.27, 0.53),
+            'dark_red_3': (0.4, 0, 0),
+            'dark_purple_3': (0.13, 0.01, 0.3),
+            'dark_magenta_3': (0.3, 0.07, 0.19),
+            'dark_cyan_3': (0.05, 0.21, 0.24),
+            'dark_orange_3': (0.5, 0.25, 0.02)
+        }
+
+        self.extra_light_colors = {
+            "light_yellow_1": (255, 217, 102),
+            "light_green_1": (147, 196, 125),
+            "light_yellow_3": (255, 242, 204),
+            "light_green_3": (217, 234, 211),
+        }
+
+        self.requests = []
+        self.service = service
+        self.sims = sims
+        self.config = config
+        self.offset_row = 0
+        self.start_of_con_in_sim = 0
+        self.start_of_arb_in_sim = 0
+
+    # find and register all unique arbiters
+
+    def assign_arbiters(self, sim):
+
+        arbiters = {}
+        count = 0
+
+        for dp in sim:
+            if isinstance(dp, ArbiterDataPoint) and dp.arbiter_id not in arbiters:
+
+                arbiters[dp.arbiter_id] = (dp.arbiter_id, count)
+                count += 1
+
+                pass_arb_Id = dp.arbiter_id
+                self.create_var_for_arb(arbiters, pass_arb_Id)
+
+        self.start_of_con_in_sim = self.offset_row
+
+        return arbiters
+
+    #find and register all consumers and assign a color to each unique one
+
+    def assign_colors_to_consumers(self, sim, arbiters):
+
+        ctcr = {}
+        count = 0
+
+        l_colors = list(self.l_c_dict.keys())
+        d_colors = list(self.d_c_dict.keys())
+
+        for dp in sim:
+            if isinstance(dp, ConsumerDataPoint) and dp.consumer_id not in ctcr:
+
+                l_color = l_colors[count]
+                d_color = d_colors[count]
+                ctcr[dp.consumer_id] = (l_color, d_color, count, dp.cycle_number)
+                count += 1
+                pass_con_Id = dp.consumer_id
+
+                self.create_var_for_con(ctcr, pass_con_Id)
+
+        return ctcr
+
+    #print the arb variables
+    #in order to do so, require to have an updated self.offset_row
+
+    def create_var_for_arb(self, arbiters, pass_arb_Id):
+        row = self.offset_row
+
+        self.requests.append(input_data_into_cell(row, 0, pass_arb_Id))
+
+        self.offset_row += 1
+
+    #print the arb variables
+    #in order to do so, require to have an updated self.offset_row
+
+    def create_var_for_con(self, ctcr, pass_con_Id):
```

Comment:
Need a better name for the function. And a better comment. Can't tell what it does.